scipy/sparse/_dok.py

```diff
@@ -344,6 +351,16 @@ def copy(self):
 
     copy.__doc__ = _spbase.copy.__doc__
 
+    @classmethod
+    def fromkeys(cls, iterable, value=1, /):
+        if isinstance(next(iter(iterable)), tuple):
```

Comment:
Sorry for missing this earlier, but: we need to create the dict first, and then figure out the shape from that.

The issue is that the `iterable` argument may be a generator, so we can't iterate over it twice. Here's a case that we should include in the test suite:

```python
dok_matrix.fromkeys((a, a+1) for a in range(3))
```

With the current approach, we end up with the right shape, but no actual entries.

To fix it, we can do:

```python
tmp = dict.fromkeys(iterable, value)
keys = tmp.keys()
# determine the shape as before, but use `keys` instead of `iterable`
# ...
result = cls(shape, dtype=type(value))
result._dict = tmp
return result
```

Reply:
You're absolutely right !